### PR TITLE
Bug fix: Measurement properties automatic padding for DynamicsBackend initialization

### PR DESCRIPTION
**Pull request by arthurostrauss**
### Summary

An indent was added in the reference code to avoid faulty behaviours when measurement_properties was added to the DynamicsBackend for the simulation to run properly. The bug was in the fact that the measurement schedule would be added, even though it had not been initialized properly (because the initialization was in a if statement with no else). Another check with a bool test was implemented to avoid loading an empty dictionary in case no measurement_properties was to be added. 

### Details and comments

Minor correction to an emerging bug

### CODE REVIEW

**Comment by DanPuzzuoli** [comment]
Thanks for this correction. There are two more things that should be done with this:
- Add a test that verifies the functioning of this code block, that would have caught this bug. It probably makes the most sense to add it as a method to `test.dynamics.backend.test_dynamics_backend.TestDynamicsBackend`. E.g. have a check where no measurement instructions exist for any qubits, and another where some exist.
- Create a release note file describing the bug fix. The process is described in the `Release Notes` section of [CONTRIBUTING.md](https://github.com/Qiskit/qiskit-dynamics/blob/main/CONTRIBUTING.md). You'll only need to fill in the bugfix section.

**Comment by DanPuzzuoli** [comment]
Another minor change that can be included in this PR is to move line 207 to before the `for` loop, as it is just redefining the same variable each time.

**Comment by arthurostrauss** [comment]
Hi, the requested information and tests, as well as the possibility to directly load the control_channels with the from_backend() method (tests are also implemented) have been added.
Let me know if additional changes are needed.

**Review by DanPuzzuoli — CHANGES_REQUESTED** [review]
For now I'm just reviewing the `control_channel_map` construction code - once that's in order we can discuss the testing/documentation.

I've added a few in-line comments. The main thing is that this code is assuming the labels for control channels are two-tuples of qubit indices. We don't want this code to assume anything about the form of the labels, as control channels are, in principle, completely arbitrary and could be labelled by anything.

Aside from what's in the in-line comments, I have the following suggestion for tightening up the code:
```
control_channel_map_backend = None
if hasattr(backend, "control_channels"):
    control_channel_map_backend = backend.control_channels
elif hasattr(backend.configuration(), "control_channels"):
    control_channel_map_backend = backend.configuration().control_channels

if bool(control_channel_map_backend):
    # extract indices from channels
    control_channel_map_backend = {label: channel[0].index for label, channel in control_channel_map_backend.items()}

   # insert code constructing control_channel_map here
```

File: qiskit_dynamics/backend/dynamics_backend.py [hunk 1]
```diff
@@ -682,6 +683,31 @@ def from_backend(
...
+        if "control_channel_map" not in options:
+            control_channel_map = {}
+            if hasattr(backend, "control_channels"):  # Look in backend.control_channels
+                control_channel_map_backend = {**{qubits: backend.control_channels[qubits][0].index
+                                                  for qubits in backend.control_channels}
+                                               }
+            elif hasattr(backend, "coupling_map"):  # Look in backend.coupling_map
+                control_channel_map_backend = {
+                    **{qubits: backend.control_channel(qubits)[0].index for qubits in backend.coupling_map}
+                }
```
[the remark this answers is not in the document]
Comment (arthurostrauss) [reply qiskit_dynamics/backend/dynamics_backend.py:696]: If the labels can be completely arbitrary, Is there a way to reduce the control_channel_map dictionary to take into account only the qubits mentioned in the subsystem_list argument of the from_backend method? 
Otherwise my last block loading only the qubits in subsystem_list will not be general in that sense. Any idea how to frame it more generally?

File: qiskit_dynamics/backend/dynamics_backend.py [hunk 2]
```diff
@@ -682,6 +683,31 @@ def from_backend(
...
+                                                  for qubits in backend.configuration().control_channels}
+                                               }
+            else:
+                control_channel_map_backend = {}
+
+            # Reduce control_channel_map to match subsystem_list
+            if bool(control_channel_map_backend):
+                for qubits in control_channel_map_backend:
+                    if qubits[0] in subsystem_list and qubits[1] in subsystem_list:
+                        control_channel_map[qubits] = control_channel_map_backend[qubits]
```
[the remark this answers is not in the document]
Comment (arthurostrauss) [reply qiskit_dynamics/backend/dynamics_backend.py:708]: Is there a logic in the arbitrary labels of those control_channels_map I could therefore use to reduce the control_channel_map to the provided subsystem_list argument? Or is it fine just loading the full one even though some of the channels will not be relevant in the frame of the simulation of the subsystem?

[the remark this answers is not in the document]
Comment (arthurostrauss) [reply qiskit_dynamics/backend/dynamics_backend.py:708]: For the reduction to the subsystem_list, you will see in the corresponding test that channels having one qubit out of two with an index out of the subsystem_list (in the tuple form (ctrl, tgt) will be appearing in the control_channel_map. Is it a desirable behavior or should we enforce control_channels that only involve labels from the subsystem_list. Acknowledging that this is the limited use case and not the generic one we discuss here.

[the remark this answers is not in the document]
Comment (DanPuzzuoli) [reply qiskit_dynamics/backend/dynamics_backend.py]: If the channel is in the hamiltonian model then it needs to be included in `control_channel_map`, regardless of what the label is or represents. Again, given that the labels can be arbitrary, we shouldn't be making decisions based on what they are.

It doesn't hurt to have the channel included in the model or in the control channel map; I think it's up to the user to be careful when restricting to subsets of qubits.

File: test/dynamics/backend/test_dynamics_backend.py [hunk 3]
```diff
@@ -550,6 +552,40 @@ def test_metadata_transfer(self):
         self.assertDictEqual(res.results[1].header.metadata, {"key1": "value1"})
 
 
+    def test_valid_measurement_properties(self):
+        """Test that DynamicsBackend instantiation always carries measurement instructions"""
+
+        # Case where no measurement instruction is added manually (which is the case for self.simple_backend)
+        instruction_schedule_map = self.backend_2q.target.instruction_schedule_map()
+        assert instruction_schedule_map.has(instruction='measure', qubits=self.simple_backend.num_qubits)
```
Comment (DanPuzzuoli) [inline test/dynamics/backend/test_dynamics_backend.py:560]: What exactly is this line checking? I'm confused by setting `qubits=self.simple_backend.num_qubits`; what happens if you pass `qubits=x` for `x` an index outside of the qubit index range?

**Review by DanPuzzuoli — CHANGES_REQUESTED** [review]
Things are looking pretty good. I'm getting way more pedantic with my comments as I think it's good aside from minor polishing.

Aside from the comments I put directly in the code, the last ones are:
- Use `self.assertTrue(x)` instead of `assert x` and `self.assertEqual(x, y)` instead of `assert x == y`, as this is the right way to do things in `unittest`.
- Make sure test method doc strings are well formed sentences. There are some periods missing.

You will also need to run automatic formatting and linting, using terminal commands `tox -e black` and `tox -e lint`.

File: test/dynamics/backend/test_dynamics_backend.py [hunk 4]
```diff
@@ -550,6 +552,40 @@ def test_metadata_transfer(self):
...
+        for q in range(self.simple_backend.num_qubits):
+            assert isinstance(instruction_schedule_map.get('measure', q).instructions[0][1], pulse.Acquire)
+            assert len(instruction_schedule_map.get('measure', q).instructions) == 1
```
[the remark this answers is not in the document]
Comment (arthurostrauss) [reply test/dynamics/backend/test_dynamics_backend.py:563]: That's correct, I removed the previous line (which I initially wanted to use over qubits= range(self.simple_backend.num_qubits)) and added it in the for loop. I have no test to check what happens if the index of the qubit provided is higher than the num_qubits argument of the backend but is that necessary?

File: test/dynamics/backend/test_dynamics_backend.py [hunk 5]
```diff
@@ -550,6 +552,40 @@ def test_metadata_transfer(self):
...
+
+        measure_properties = {(0,): InstructionProperties(calibration=meas_sched0),
+                              (1,): InstructionProperties(calibration=meas_sched1)}
+        target = Target()
+        target.add_instruction(Measure(), measure_properties)
+        custom_meas_backend = DynamicsBackend(solver=self.solver_2q, target=target, subsystem_dims=[2, 2])
+        instruction_schedule_map = custom_meas_backend.target.instruction_schedule_map()
+        for q in range(self.simple_backend.num_qubits):
+            assert isinstance(instruction_schedule_map.get('measure', q).instructions[0][1], pulse.Acquire)
+            assert len(instruction_schedule_map.get('measure', q).instructions) == 1
```
[the remark this answers is not in the document]
Comment (arthurostrauss) [reply test/dynamics/backend/test_dynamics_backend.py:585]: Not sure how to replace it by the has method, I added it on top of what I already have written. The point here is to ensure that the instruction stored in the target now is the custom one that has overridden the previous default one (checking that by the custom_duration argument value)

File: releasenotes/notes/measurement_property_bug_fix-12461088823a943c.yaml [hunk 6]
```diff
@@ -1,19 +1,11 @@
...
+    `control_channel_map` option based on the supplied backend if the user does not supply one.
 
 fixes:
   - |
-    A bug regarding the setting of the measurement properties at the initialisation of the
-    DynamicsBackend instance was always resetting the measurement instruction to a default setting 
-    even though there was already a measurement instruction stored in the provided Target object. This bug is now fixed
-    and the measurement instruction is added only if the designated qubit in the provided Target does not have a measurement instruction.
+    A bug in :meth:`DynamicsBackend.__init__` causing existing measurement instructions for a user-supplied
+    :class:`Target` has been fixed.
```
Comment (DanPuzzuoli) [inline releasenotes/notes/measurement_property_bug_fix-12461088823a943c.yaml:10]: Apologies this sentence is incomplete :).

```suggestion
    A bug in :meth:`DynamicsBackend.__init__` causing existing measurement instructions for a user-supplied
    :class:`Target` to be overwritten has been fixed.
```

File: test/dynamics/backend/test_dynamics_backend.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (DanPuzzuoli) [reply test/dynamics/backend/test_dynamics_backend.py]: No I don't think so.

File: test/dynamics/backend/test_dynamics_backend.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (DanPuzzuoli) [reply test/dynamics/backend/test_dynamics_backend.py]: Yeah what you're checking makes sense. I was specifically referring to the line:
```
self.assertEqual(instruction_schedule_map.get("measure", q).instructions, 1)
```
Is this the same as checking:
```
self.assertTrue(instruction_schedule_map.has("measure", q))
```
?

File: test/dynamics/backend/test_dynamics_backend.py (hunk 3 above)
Comment (DanPuzzuoli) [reply test/dynamics/backend/test_dynamics_backend.py]: I see this line has been removed.

File: test/dynamics/backend/test_dynamics_backend.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (arthurostrauss) [reply test/dynamics/backend/test_dynamics_backend.py]: So the line actually checks the length of the instructions list. It's actually:
`self.assertEqual(len(instruction_schedule_map.get("measure", q).instructions), 1)`

So in essence it just checks that the measure instruction is only composed of one Acquire Pulse.

**Review by DanPuzzuoli — APPROVED** [review]
LGTM, thanks for the contribution!
